bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/crypto/EncryptionHandler.java

```diff
@@ -0,0 +1,307 @@
+ * Encrypt and Decrypt methods use the expandedKey field to do their job.
+ *
+ * The first packet sent to Paradox is the IP150 password as bytes, extended to 32 bytes with 0xEE.
+ * The first response contains the key that will be used for the rest of communication.
+ *
+ * Most of the coding is copy from Python / rewrite in Java from second link of jpbaracca's PAI repository. Probably
+ * some of the variables can be named better but I don't understand this code in it's full scope so I preferred to keep
+ * it as it is.
+ *
+ * @author Konstantin Polihronov - Initial contribution
+ *
+ * @see <a href=https://www.samiam.org/key-schedule.html>Sam Trendholme's page about AES</a>
+ * @see <a href=https://github.com/ParadoxAlarmInterface/pai>Github of jpbaracca's work - ParadoxAlarmInterface in
+ *      python</a>
+ */
+public class EncryptionHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EncryptionHandler.class);
+
+    private static final int KEY_ARRAY_LENGTH = 32;
+    private static final int TABLE_SIZE = 256;
+    private static final int KEY_LENGTH = 240;
+    private static final int PAYLOAD_RATE_LENGTH = 16;
+
+    private static EncryptionHandler instance;
+
+    private int[] lTable = new int[TABLE_SIZE];
+    private int[] aTable = new int[TABLE_SIZE];
+
+    private int[] expandedKey = new int[KEY_LENGTH];
+
+    private EncryptionHandler() {
+        generateTables();
+    }
+
+    public static EncryptionHandler getInstance() {
+        if (instance == null) {
+            synchronized (EncryptionHandler.class) {
+                instance = new EncryptionHandler();
+            }
+        }
+        return instance;
+    }
+
+    public byte[] encrypt(byte[] payload) {
+        if (payload == null) {
+            throw new IllegalArgumentException("Payload should not be null or empty !");
+        }
+
+        if (payload.length % 16 != 0) {
+            payload = ParadoxUtil.extendArray(payload, PAYLOAD_RATE_LENGTH);
+            printArray("Array had to be extended:", payload);
+            logger.trace("New payload length={}", payload.length);
+        }
+
+        int[] payloadAsIntArray = ParadoxUtil.copyArray(payload);
+
+        final int[] s = EncryptionHandlerConstants.S;
+        byte[] result = new byte[0];
+        for (int i = 0, rounds = 14; i < payloadAsIntArray.length / 16; i++) {
+            int[] a = Arrays.copyOfRange(payloadAsIntArray, i * 16, (i + 1) * 16);
+            keyAddition(a, Arrays.copyOfRange(expandedKey, 0, 16));
+
+            for (int r = 1; r < rounds; r++) {
+                sBox(a, s);
+                shiftRow(a, 0);
+                mixColumn(a);
+                keyAddition(a, Arrays.copyOfRange(expandedKey, r * 16, (r + 1) * 16));
+            }
+            sBox(a, s);
+            shiftRow(a, 0);
+            keyAddition(a, Arrays.copyOfRange(expandedKey, 16 * rounds, (rounds + 1) * 16));
```

Comment:
```suggestion
            for (int r = 1; r <= rounds; r++) {
                sBox(a, s);
                shiftRow(a, 0);
                if(r != rounds){
                    mixColumn(a);
                }
                keyAddition(a, Arrays.copyOfRange(expandedKey, r * 16, (r + 1) * 16));
            }
```

Reply:
Thanks for pointing that out. Seeing it now it seems obvious :(
Will fix it asap.